Return local cluster IP for headless services

mikenorgate opened the pull request:
This fixes issue #656 where DNS resolution of a headless services 
does not return the local IP for pods within the local cluster

dfarrell07 commented:
Thank you for the PR @mikenorgate!

vthapar commented:
@mikenorgate Thanks for working on this. Have you tested it locally? Just looking at code I'd expect it to work for local clusters, but break for remote.

mikenorgate commented:
> @mikenorgate Thanks for working on this. Have you tested it locally? Just looking at code I'd expect it to work for local clusters, but break for remote.

@vthapar I've been testing it in local KinD clusters at it looks to be behaving correctly. Can you explain a bit more of what you mean by break for remote clusters?

vthapar commented on pkg/agent/controller/endpoint.go:
```diff
@@ -274,7 +275,7 @@ func allAddressesIPv6(addresses []corev1.EndpointAddress) bool {
 
 func (e *EndpointController) getIP(address *corev1.EndpointAddress) string {
-	if e.isHeadless && e.globalIngressIPCache != nil {
+	if e.clusterID != e.localClusterID && e.isHeadless && e.globalIngressIPCache != nil {
```
This will work for local cluster but break remote clusters. This is part of code where we create EndpointSlices from service's endpoints to sync to remote clusters. ClusterId check here is irrelevant as we will only hit this code for local services on lighthouse agent.

The files you need to modify and handle clusterIDs are in pkg/endpointslice/

mikenorgate replied:
I understand what you mean now, I'll go back and try to come up with another fix

mikenorgate commented:
@vthapar I've changed my approach and would appreciate your feedback. I still have a couple of tests to fix up but wanted to make sure I was on the correct path now

vthapar commented on pkg/endpointslice/map.go:
```diff
@@ -111,6 +117,20 @@ func (m *Map) Put(es *discovery.EndpointSlice) {
+			}).String(),
+		})
+		if err != nil || len(localSlices.Items) != 1 {
+			klog.Warningf("Unable to find local endpoint slice for service")
```
This doesn't need to be warning but debug. Normal operation will have lots of cases where we only have remote services. This will make our logs fill up with warning logs which don't add any value. A debug level will work better as then it can be used, by increasing log level, to troubleshoot specific failure.

vthapar commented on test/e2e/discovery/headless_services.go:
```diff
@@ -127,7 +127,12 @@ func RunHeadlessDiscoveryLocalAndRemoteTest(f *lhframework.Framework) {
-	ipListA, hostNameListA := f.GetPodIPs(framework.ClusterA, nginxHeadlessClusterA)
+	_, hostNameListA := f.GetPodIPs(framework.ClusterA, nginxHeadlessClusterA)
+
+	var ipListA []string
```
How about modifying GetPodIPs itself to return podIPs if cluster is local? Maybe add an `isLocal` flag? That way complete logic sits in GetPodIPs and test code stays the same and easy to read.

vthapar commented on test/e2e/discovery/headless_services.go:
```diff
@@ -217,9 +222,14 @@ func RunHeadlessDiscoveryClusterNameTest(f *lhframework.Framework) {
+	_, hostNameListA := f.GetPodIPs(framework.ClusterA, nginxHeadlessClusterA)
 	ipListClusterB, hostNameListB := f.GetPodIPs(framework.ClusterB, nginxHeadlessClusterB)
 
+	var ipListClusterA []string
```
same as above

vthapar commented on pkg/endpointslice/map.go:
```diff
@@ -111,6 +117,20 @@ func (m *Map) Put(es *discovery.EndpointSlice) {
+				"kubernetes.io/service-name": es.Labels[constants.MCSLabelServiceName],
+			}).String(),
+		})
+		if err != nil || len(localSlices.Items) != 1 {
```
Need to add a retry in case error is due to timeout. Those are not that uncommon if Kube API server is under heavy traffic.

vthapar commented on pkg/endpointslice/map.go:
```diff
@@ -142,7 +162,18 @@ func (m *Map) Put(es *discovery.EndpointSlice) {
+		addresses := endpoint.Addresses
+		if localSlice != nil {
+			for _, localEndpoint := range localSlice.Endpoints {
+				if *localEndpoint.NodeName == *endpoint.NodeName &&
```
Not sure if we need this check. These would be same for non-globalnet while for globalnet localslice we fetched here will be more up-to-date and recent than the one we got from event notification.

mikenorgate replied:
I found that the `localSlice` didn't include the hostname which is why I attempt to map back to the globalnet endpoint slice

vthapar commented on pkg/endpointslice/map.go:
```diff
@@ -111,6 +117,20 @@ func (m *Map) Put(es *discovery.EndpointSlice) {
 	}
 
+	var localSlice *discovery.EndpointSlice
+	if cluster == m.localClusterID {
```
We should also add globalnet flag and do this only for globalnet. For non globalnet this is irrelevant and will just add extra calls to Kube API server.

This probably means adding another env to lighthouse-coredns deployment. It can come in a later patch and I'd recommend adding a TODO here and creating an issue to track it.

vthapar reviewed (changes requested):
Overall approach is good and exactly what I had in mind. Some comments regarding error handling and possibe optimizations.

Would be good to create a separate E2E test for HeadlessLocal only where we test change in number of endpoints by increasing and decreasing replica set size.

vthapar commented:
@mikenorgate I've pushed https://github.com/submariner-io/lighthouse/pull/762 It should make things simpler for you.

dfarrell07 commented:
This needs a manual rebase before re-running CI.

dfarrell07 commented:
Thanks for the rebase @mikenorgate, I ran CI and it's showing some minor issues.

Can you re-review @vthapar?

vthapar commented on coredns/endpointslice/map.go:
```diff
@@ -111,6 +119,35 @@ func (m *Map) Put(es *discovery.EndpointSlice) {
 		return
 	}
 
+	// TODO: Only do this look up if globalnet enabled to reduce unnessary call to Kube API
```
Can you create an issue to track this TODO item?

mikenorgate replied:
Created #775

vthapar reviewed (approved):
Thanks for all the effort on this, looks good. Tested locally and confirmed to be working.

dfarrell07 commented:
This has a number of CI failures (go linting/unit, commit message/DCO) that need to be fixed before it can be merged.

I'd suggest going ahead and squashing all the commits together into (likely 1) logically-atomic commit. More details:

https://submariner.io/development/code-review/#squashamend-commits-into-discrete-steps

vthapar commented:
@mikenorgate Can you resolve the linter and UT failures?

mikenorgate commented:
@vthapar I've fixed the linter errors but I can't see what the reason for the unit test failure is, all the tests are passing and there doesn't seem to be an error message in the output

tpantelis commented:
> @vthapar I've fixed the linter errors but I can't see what the reason for the unit test failure is, all the tests are passing and there doesn't seem to be an error message in the output

The failure is due to [compilation errors](https://github.com/submariner-io/lighthouse/runs/6881666688?check_suite_focus=true#step:4:921) in _coredns/plugin/handler_test.go_.

tpantelis commented on coredns/endpointslice/map.go:
```diff
@@ -86,13 +93,14 @@ func (m *Map) GetDNSRecords(hostname, cluster, namespace, name string, checkClus
 }
 
-func (m *Map) Put(es *discovery.EndpointSlice) {
+func (m *Map) Put(es *discovery.EndpointSlice, localClient v1.DiscoveryV1Interface) {
```
I think it's cleaner to pass `localClient` into the `NewMap` ctor and store in `Map`.

dfarrell07 commented:
While you're making wonderful contributions to Submariner @mikenorgate, maybe you'd like to add your company to the ADOPTERS.md file? We have a CNCF review coming up soon and it would be great to make it reflect reality a bit more.

https://github.com/submariner-io/submariner/blob/devel/ADOPTERS.md

mikenorgate commented:
I've made the requested changes and fixed the unit tests

tpantelis commented on coredns/endpointslice/map.go:
```diff
@@ -86,9 +95,11 @@ func (m *Map) GetDNSRecords(hostname, cluster, namespace, name string, checkClus
 }
 
-func NewMap() *Map {
+func NewMap(localClusterID string, kubeConfig *rest.Config) *Map {
```
It's better to pass in a `kubernetes.Interface` - we avoid passing `rest.Config` b/c there's no mock/fake implementation for unit tests.

```suggestion
func NewMap(localClusterID string, kubeClient kubernetes.Interface) *Map {
```

tpantelis commented on coredns/endpointslice/map.go:
```diff
@@ -142,7 +188,18 @@ func (m *Map) Put(es *discovery.EndpointSlice) {
+		addresses := endpoint.Addresses
+		if localSlice != nil {
+			for _, localEndpoint := range localSlice.Endpoints {
+				if *localEndpoint.NodeName == *endpoint.NodeName &&
```
Is it certain that either `NodeName` won't be nil?

mikenorgate replied:
The node name of `endpoint` wont be nil as it is set here https://github.com/submariner-io/lighthouse/blob/4a48ab0ef6769397760c6f268c04108c008a4e20/pkg/agent/controller/endpoint.go#L242 
The local endpoint could potentially have a nil node name but if that happens we not going to be able to tell if that endpoint is the one needed so should probably ignore it

tpantelis replied:
The `NodeName` pointer for `endpoint` is assigned from `address` which theoretically could be be nil. We just want to avoid a crash if somehow either is nil so, if we're not certain, then we should check for nil and perhaps log a warning.

mikenorgate replied:
I've added a check to make sure they're not nil

tpantelis commented on coredns/endpointslice/map.go:
```diff
@@ -111,6 +122,41 @@ func (m *Map) Put(es *discovery.EndpointSlice) {
+			})
+			if err != nil && !apierrors.IsNotFound(err) {
+				return err
+			} else if err != nil || len(localSlices.Items) != 1 {
```
We don't really need to check for `IsNotFound` error - it would be reported on L152 below anyway. Also for the `List` operation, it doesn't return `IsNotFound` if there's no items found - it only returns `IsNotFound` if the resource type isn't found which shouldn't happen.

```suggestion
			if err != nil {
				return err
			} 
			
			if len(localSlices.Items) != 1 {
```

tpantelis commented on coredns/endpointslice/map.go:
```diff
@@ -111,6 +122,41 @@ func (m *Map) Put(es *discovery.EndpointSlice) {
+			if err != nil && !apierrors.IsNotFound(err) {
+				return err
+			} else if err != nil || len(localSlices.Items) != 1 {
+				klog.V(log.DEBUG).Infof("Unable to find local endpoint slice for service %s/%s", es.Labels[constants.LabelSourceNamespace], es.Labels[constants.MCSLabelServiceName])
+				return nil
```
For the case where there's more than one `EndpointSlice` found, this log message isn't actually  accurate. Also we log to debug here as opposed to error on L152. However, do we need to fail in this case, ie can't we just take the first one? What would be the case where there might be more than one?

For the case where the returned list is empty, we could return a `NotFound` error and let if fall through to L152, ie 

```Go
  apierrors.NewNotFound(schema.GroupResource{
		Group:    discovery.SchemeGroupVersion.Group,
		Resource: "endpointslices",
	}, "")
```

mikenorgate replied:
I think taking the first one should be fine, if there ever is more than one they should all point to the same set of addresses

tpantelis commented on coredns/endpointslice/map.go:
```diff
@@ -111,6 +122,41 @@ func (m *Map) Put(es *discovery.EndpointSlice) {
+			klog.Warningf("Error creating client set %#v", err)
+			return
+		}
+
```
It would be nice to add unit tests to cover this new block of code but we can do that in another PR.

Commit:
Return local cluster IP for headless services

This fixes issue submariner-io#656 where DNS resolution of a headless service
does not return the local IP for pods within the local cluster

Signed-off-by: Mike Norgate <mike.norgate@mythical.games>

dfarrell07 commented:
It's weird there isn't a rebase button or a note about conflict on the PR, but it can't be merged because it's "out-of-date with the base branch".

vthapar commented:
@mikenorgate Can you rebase with latest devel so we can get this merged? Pulling this locally I can see it is missing a rebase.

maayanf24 commented:
@mikenorgate - can you please fix the lint errors?

mikenorgate commented:
> @mikenorgate - can you please fix the lint errors?

I think it just needs to be re-run, looking at the error it seems to be a networking issue
